### PR TITLE
*: incorrect error message of inserting enum & set (#19380)

### LINKED ISSUE #19229: Incorrect error message of inserting enum & set
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```mysql
CREATE TABLE enumt (type enum('a', 'b') );
insert into enumt values('xxx');
insert into enumt values('-1');

CREATE TABLE sett (type set('a', 'b') );
insert into sett values('xxx');
insert into sett values('-1');
```

### 2. What did you expect to see? (Required)

```mysql
mysql> CREATE TABLE enumt (type enum('a', 'b') );
Query OK, 0 rows affected (0.01 sec)

mysql> insert into enumt values('xxx');
ERROR 1265 (01000): Data truncated for column 'type' at row 1
mysql> insert into enumt values('-1');
ERROR 1265 (01000): Data truncated for column 'type' at row 1

mysql> CREATE TABLE sett (type set('a', 'b') );
Query OK, 0 rows affected (0.00 sec)

mysql> insert into sett values('xxx');
ERROR 1265 (01000): Data truncated for column 'type' at row 1
mysql> insert into sett values('-1');
ERROR 1265 (01000): Data truncated for column 'type' at row 1
```

### 3. What did you see instead (Required)

```mysql
mysql> CREATE TABLE enumt (type enum('a', 'b') );
Query OK, 0 rows affected (0.01 sec)

mysql> insert into enumt values('xxx');
ERROR 1366 (HY000): Incorrect enum value: 'xxx' for column 'type' at row 1
mysql> insert into enumt values('-1');
ERROR 1366 (HY000): Incorrect enum value: '-1' for column 'type' at row 1

mysql> CREATE TABLE sett (type set('a', 'b') );
Query OK, 0 rows affected (0.01 sec)

mysql> insert into sett values('xxx');
ERROR 1105 (HY000): cannot convert datum from char to type set.
mysql> insert into sett values('-1');
ERROR 1105 (HY000): cannot convert datum from char to type set.
```

### 4. What is your TiDB version? (Required)

v4.0.4

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #19380 to release-4.0

---

### What problem does this PR solve?

Issue Number: close #19229 

Problem Summary:

### What is changed and how it works?

What's Changed: Change error code and error info when insert value.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- fix incorrect error message of inserting enum & set

### CODE REVIEW

File: table/column.go [hunk 1]
```diff
@@ -181,7 +181,11 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, r
 	if types.ErrOverflow.Equal(err) && returnOverflow {
 		return casted, err
 	}
+<<<<<<< HEAD
```
Comment (qw4990) [inline table/column.go:184]: Please fix the conflict @Patrick0308 .

Comment (Patrick0308) [reply table/column.go:184]: Do I need create a new pr to fix this problem?

Comment (Patrick0308) [reply table/column.go:184]: What should I do?

**Review by XuHuaiyu — COMMENTED** [review]
CI failed, you may resolve the conflicts and file a new PR

**Comment by Patrick0308** [comment]
@XuHuaiyu @qw4990 I fix conflict in #19989.

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @qw4990

File: table/column.go [hunk 2]
```diff
@@ -181,7 +181,7 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo, r
 	if types.ErrOverflow.Equal(err) && returnOverflow {
 		return casted, err
 	}
-	if types.ErrTruncated.Equal(err) {
+	if err != nil && types.ErrTruncated.Equal(err) && col.Tp != mysql.TypeSet && col.Tp != mysql.TypeEnum {
```
Comment (Patrick0308) [inline table/column.go:184]: We don't need `err!=nil` because in `types.ErrTruncated.Equal(err)` fix nil problem.

Comment (Patrick0308) [reply table/column.go:184]: @XuHuaiyu

**Review by djshow832 — COMMENTED** [review]
LGTM

**Review by lzmhhh123 — APPROVED** [review]

**Pull request merged**